src/Plugin/Block/ChannelSearchBlock.php

```diff
@@ -0,0 +1,232 @@
+        'view' => $channel_view,
+        'display' => &$channel_view->display_handler->display,
+        'rerender' => NULL,
+        'parent-source' => self::class,
+      ])
+      ->setMethod('get')
+      ->setAlwaysProcess()
+      ->disableRedirect();
+
+    $views_exposed_filter_form = $this->formBuilder->buildForm(ViewsExposedForm::class, $form_state);
+    return $views_exposed_filter_form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCacheTags() {
+
+    $host_node = $this->getContextValue('node');
+    $channel_view = Views::getView(self::CHANNEL_VIEW);
+
+    $cache_tags = Cache::mergeTags(parent::getCacheTags(), $host_node->getCacheTags());
+    $cache_tags = Cache::mergeTags($cache_tags, $channel_view->getCacheTags());
+
+    if ($this->isDirEntry($host_node)) {
+      $cache_tags_for_dir_channels = $this->getCacheTagsForDirChannels($host_node);
+      $cache_tags = Cache::mergeTags($cache_tags, $cache_tags_for_dir_channels);
+    }
+
+    return $cache_tags;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCacheContexts() {
+
+    return Cache::mergeContexts(parent::getCacheContexts(), ['route']);
+  }
+
+  /**
+   * Is this node a Directory entry?
+   */
+  protected function isDirEntry(NodeInterface $node): bool {
+
+    return $node->hasField(self::CHANNEL_SELECTION_FIELD);
+  }
+
+  /**
+   * Cache tags for Directory channel nodes.
+   *
+   * The given Directory entry belongs to these Directory channel nodes.
+   */
+  protected function getCacheTagsForDirChannels(NodeInterface $directory_entry): array {
+
+    $directory_channels = array_filter(array_map(function (EntityReferenceItem $ref_item) {
+      return $ref_item->entity;
+    }, iterator_to_array($directory_entry->{self::CHANNEL_SELECTION_FIELD})));
+
+    $cache_tags = array_reduce(
+      $directory_channels,
+      function (array $carry, NodeInterface $directory_channel): array {
+        return Cache::mergeTags($carry, $directory_channel->getCacheTags());
+      }, []
+    );
+
+    return $cache_tags;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity.repository'),
+      $container->get('form_builder')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, $entity_repo, $form_builder) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->entityRepository = $entity_repo;
+    $this->formBuilder = $form_builder;
+  }
+
+  /**
+   * Entity repository service.
+   *
+   * @var \Drupal\Core\Entity\EntityRepositoryInterface
+   */
+  protected $entityRepository;
+
+  /**
+   * Form builder service.
+   *
+   * @var \Drupal\Core\Form\FormBuilderInterface
+   */
+  protected $formBuilder;
```

Comment:
Very minor issue, but most classes are ordered as:
1. Class static variables
2. Class variables
4. Static methods
5. Constructor
6. Class methods

Not too bothered if you disagree :wink:

Reply:
Hi Steve,
You are right that most classes follow this format.  In the absence of an agreed up on convention, I place the public methods first and the class variables last.  IIRC I picked up this convention from Pete Goodliffe's [Code craft](https://www.goodreads.com/book/show/44875.Code_Craft).  For the ChannelSearchBlock plugin class, the most important public method is ChannelSearchBlock::build().  Hence it's at the top.  The idea is, if you have to work on that block, you are most likely interested in its build() method.  So you better find it as early as possible.  I am happy to change the order.  Just let me know.

FAO @stephen-cox

Reply:
@Adnan-cds   Drupal core has all variables at the top followed by constructor, although static methods seem to be mixid in. Here's an example: https://git.drupalcode.org/project/drupal/-/blob/9.0.x/core/lib/Drupal/Core/Entity/ContentEntityBase.php

It's not a major thing and it sounds like this is holding other work up so happy to approve.

Reply:
Thanks Steve :)